### PR TITLE
libuv-devel: new port (version 1.47.0)

### PR DESCRIPTION
**Pull request by herbygillot**
###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with ~`sudo port -vst install`~ `sudo port -d destroot`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by barracuda156** [comment]
To begin with:
```
In file included from src/fs-poll.c:23:
src/uv-common.h:41:24: error: stdatomic.h: No such file or directory
In file included from src/thread-common.c:23:
src/uv-common.h:41:24: error: stdatomic.h: No such file or directory
In file included from src/random.c:23:
src/uv-common.h:41:24: error: stdatomic.h: No such file or directory
In file included from src/inet.c:23:
src/uv-common.h:41:24: error: stdatomic.h: No such file or directory
make: *** [src/libuv_la-thread-common.lo] Error 1
make: *** Waiting for unfinished jobs....
make: *** [src/libuv_la-inet.lo] Error 1
make: *** [src/libuv_la-random.lo] Error 1
make: *** [src/libuv_la-fs-poll.lo] Error 1
```
I think this also gonna affect some Intel systems. I.e. newer compilers should be used on old systems.

**Comment by barracuda156** [comment]
@herbygillot Then this fails, so our patches *are* needed:
```
libtool: compile:  /opt/local/bin/gcc-mp-13 -DPACKAGE_NAME=\"libuv\" -DPACKAGE_TARNAME=\"libuv\" -DPACKAGE_VERSION=\"1.46.0\" "-DPACKAGE_STRING=\"libuv 1.46.0\"" -DPACKAGE_BUGREPORT=\"https://github.com/libuv/libuv/issues\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libuv\" -DVERSION=\"1.46.0\" -DSUPPORT_ATTRIBUTE_VISIBILITY_DEFAULT=1 -DSUPPORT_FLAG_VISIBILITY=1 -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_PTHREAD_PRIO_INHERIT=1 -I. -I./include -I./src -I./src/unix -isystem/opt/local/include/LegacySupport -I/opt/local/include -D__DARWIN_64_BIT_INO_T -fno-strict-aliasing -g -std=gnu89 -Wall -Wextra -Wno-long-long -Wno-unused-parameter -Wstrict-prototypes -D_DARWIN_USE_64_BIT_INODE=1 -D_DARWIN_UNLIMITED_SELECT=1 -pipe -Os -arch ppc -fvisibility=hidden -pthread -MT src/unix/libuv_la-bsd-ifaddrs.lo -MD -MP -MF src/unix/.deps/libuv_la-bsd-ifaddrs.Tpo -c src/unix/bsd-ifaddrs.c  -fno-common -DPIC -o src/unix/.libs/libuv_la-bsd-ifaddrs.o
src/unix/udp.c: In function 'uv__udp_set_source_membership4':
src/unix/udp.c:901:25: error: storage size of 'mreq' isn't known
  901 |   struct ip_mreq_source mreq;
      |                         ^~~~
src/unix/udp.c:923:15: error: 'IP_ADD_SOURCE_MEMBERSHIP' undeclared (first use in this function); did you mean 'IP_ADD_MEMBERSHIP'?
  923 |     optname = IP_ADD_SOURCE_MEMBERSHIP;
      |               ^~~~~~~~~~~~~~~~~~~~~~~~
      |               IP_ADD_MEMBERSHIP
src/unix/udp.c:923:15: note: each undeclared identifier is reported only once for each function it appears in
src/unix/udp.c:925:15: error: 'IP_DROP_SOURCE_MEMBERSHIP' undeclared (first use in this function); did you mean 'IP_DROP_MEMBERSHIP'?
  925 |     optname = IP_DROP_SOURCE_MEMBERSHIP;
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~
      |               IP_DROP_MEMBERSHIP
src/unix/udp.c:901:25: warning: unused variable 'mreq' [-Wunused-variable]
  901 |   struct ip_mreq_source mreq;
      |                         ^~~~
src/unix/udp.c: In function 'uv__udp_set_source_membership6':
src/unix/udp.c:946:27: error: storage size of 'mreq' isn't known
  946 |   struct group_source_req mreq;
      |                           ^~~~
mv -f src/unix/.deps/libuv_la-thread.Tpo src/unix/.deps/libuv_la-thread.Plo
src/unix/udp.c:972:15: error: 'MCAST_JOIN_SOURCE_GROUP' undeclared (first use in this function)
  972 |     optname = MCAST_JOIN_SOURCE_GROUP;
      |               ^~~~~~~~~~~~~~~~~~~~~~~
/bin/sh ./libtool  --tag=CC   --mode=compile /opt/local/bin/gcc-mp-13 -DPACKAGE_NAME=\"libuv\" -DPACKAGE_TARNAME=\"libuv\" -DPACKAGE_VERSION=\"1.46.0\" -DPACKAGE_STRING=\"libuv\ 1.46.0\" -DPACKAGE_BUGREPORT=\"https://github.com/libuv/libuv/issues\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libuv\" -DVERSION=\"1.46.0\" -DSUPPORT_ATTRIBUTE_VISIBILITY_DEFAULT=1 -DSUPPORT_FLAG_VISIBILITY=1 -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_PTHREAD_PRIO_INHERIT=1 -I.  -I./include -I./src  -I./src/unix -isystem/opt/local/include/LegacySupport -I/opt/local/include -D__DARWIN_64_BIT_INO_T -fno-strict-aliasing -g -std=gnu89 -Wall -Wextra -Wno-long-long -Wno-unused-parameter -Wstrict-prototypes      -D_DARWIN_USE_64_BIT_INODE=1 -D_DARWIN_UNLIMITED_SELECT=1      -pipe -Os -arch ppc -fvisibility=hidden -pthread -MT src/unix/libuv_la-darwin-proctitle.lo -MD -MP -MF src/unix/.deps/libuv_la-darwin-proctitle.Tpo -c -o src/unix/libuv_la-darwin-proctitle.lo `test -f 'src/unix/darwin-proctitle.c' || echo './'`src/unix/darwin-proctitle.c
src/unix/udp.c:974:15: error: 'MCAST_LEAVE_SOURCE_GROUP' undeclared (first use in this function)
  974 |     optname = MCAST_LEAVE_SOURCE_GROUP;
      |               ^~~~~~~~~~~~~~~~~~~~~~~~
src/unix/udp.c:946:27: warning: unused variable 'mreq' [-Wunused-variable]
  946 |   struct group_source_req mreq;
      |                           ^~~~
mv -f src/unix/.deps/libuv_la-tty.Tpo src/unix/.deps/libuv_la-tty.Plo
make: *** [src/unix/libuv_la-udp.lo] Error 1
make: *** Waiting for unfinished jobs....
```

**Comment by barracuda156** [comment]
@herbygillot Just pick this: https://github.com/macports/macports-ports/blob/master/devel/uvw/files/patch-libuv-1.46.diff

**Comment by herbygillot** [comment]
Thank you for testing, @barracuda156 - I've added the patch you referenced.

**Comment by barracuda156** [comment]
@herbygillot We still need to set a C standard (due to stdatomics). It won’t build with old Xcode gccs and, I think, some old clangs too.

**Comment by herbygillot** [comment]
@barracuda156 you can add an additional commit to this branch including everything you think is needed, and we can squash/clean up when it's good to be merged.

**Comment by barracuda156** [comment]
@herbygillot Not sure I technically will be allowed to commit here, but I can try once on a laptop.

**Comment by barracuda156** [comment]
@herbygillot Ok, how do I add this here? https://github.com/macports/macports-ports/commit/59bf8c252c7b701b2686b5ed7bcf419b863495a8
Tree: https://github.com/barracuda156/macports-ports/tree/libuv-devel

**Comment by casr** [comment]
Would it be worth doing `-devel` as a subport similar to how [`tmux`](https://github.com/macports/macports-ports/blob/88daea7ca7187858b5d67d847cc4bb333a16fe5d/sysutils/tmux/Portfile#L10-L20) does it? It'd probably reduce a lot of the repetition regarding patches, dependencies, etc

**Comment by herbygillot** [comment]
> Would it be worth doing `-devel` as a subport similar to how [`tmux`](https://github.com/macports/macports-ports/blob/88daea7ca7187858b5d67d847cc4bb333a16fe5d/sysutils/tmux/Portfile#L10-L20) does it? It'd probably reduce a lot of the repetition regarding patches, dependencies, etc

Yes, this makes a lot more sense. I've now switched `libuv-devel` over to be a subport.

**Comment by herbygillot** [comment]
> @herbygillot Ok, how do I add this here? [59bf8c2](https://github.com/macports/macports-ports/commit/59bf8c252c7b701b2686b5ed7bcf419b863495a8) Tree: https://github.com/barracuda156/macports-ports/tree/libuv-devel

@barracuda156, I could not add your exact commit since I changed from using a separate Portfile to making it a subport, but I added your changes.

To answer your question on adding commits to a PR:

1. First you'll need to add the PR branch as a remote.  This PR's branch is [herbygillot:libuv-devel](https://github.com/herbygillot/macports-ports/tree/libuv-devel), so you add it as another remote:
  `git remote add herbygillot git@github.com:herbygillot/macports-ports.git`
2. Fetch it: `git fetch herbygillot libuv-devel`
3. Check it out: `git checkout -b herby-libuv-devel herbygillot/libuv-devel`
4. Make your changes and add additional commits.
5. It's possible you might run into permissions at this point, but you then push your changes back the PR branch:
  `git push herbygillot libuv-devel:herby-libuv-devel`

Can you test this branch when you get a chance?

**Comment by barracuda156** [comment]
@herbygillot Please do not merge this yet:
```
Undefined symbols:
  "_preadv", referenced from:
      _uv__fs_work in libuv_la-fs.o
  "_pwritev", referenced from:
      _uv__fs_work in libuv_la-fs.o
ld: symbol(s) not found
collect2: error: ld returned 1 exit status
```

**Comment by barracuda156** [comment]
@catap @macportsraf @kencu Could we add these to `legacysupport`?

`folly` has them in portability code:
https://github.com/facebook/folly/commit/f464accb69e96176e5a313da5d24d76ee5297ca2
`wrapPositional` is here: https://github.com/facebook/folly/blob/26a2c0267a0dbf9aa461106fe60aa44e74ce3b01/folly/portability/Unistd.cpp#L60-L86

**Comment by herbygillot** [comment]
Thank you for tearing, @barracuda156

**Comment by herbygillot** [comment]
> @herbygillot Please do not merge this yet:
> ```
> Undefined symbols:
>   "_preadv", referenced from:
>       _uv__fs_work in libuv_la-fs.o
>   "_pwritev", referenced from:
>       _uv__fs_work in libuv_la-fs.o
> ld: symbol(s) not found
> collect2: error: ld returned 1 exit status
> ```
> Apparently, these are missing up to 10.13 (!). `libuv` upstream is doing strange things.

Just fyi @kencu

**Comment by catap** [comment]
@barracuda156 one day — defently

**Comment by barracuda156** [comment]
@catap Ok, let me try to make usage of those conditional for now – we can always drop that once they are, hopefully, added to `legacysupport`.

**Comment by catap** [comment]
@barracuda156 I suggest to start from implementing it on legacy support

**Comment by barracuda156** [comment]
@catap I am afraid this is not something what I can do myself (I also got no environment to test such things – which obviously should be tested not just on PowerPC).

**Comment by kencu** [comment]
libuv upstream has officially dropped all support for 10.15 and less, they are not accepting any patches or other workarounds into their codebase for older systems support, and they are actively stripping out whatever workarounds remain in the codebase.

**Comment by kencu** [comment]
whether what libuv needs can replaced by legacysupport, or otherwise worked around in a realistically straightforward manner with some maintainable patches remains to be seen.

for now, it seems highly likely that libuv will have to be pegged back on 10.15 and less, and then updated from time-to-time depending on what can be done and by people with involved systems who want to make the effort.

IMHO the most realistic thing to do would be to generate a libuv-legacy port and have libuv depend on that on older systems, much like libsdl2 does now.

**Comment by barracuda156** [comment]
@kencu It is quite possible, but I guess not now, yet – or at least not for the sake of `preadv`.

**Comment by barracuda156** [comment]
@herbygillot I think it does not allow me to push to your branch directly, but here is the fix: https://github.com/macports/macports-ports/commit/39f38737566e0fa87f0ed4897e467ab96d7ab9ab

**Comment by herbygillot** [comment]
> @herbygillot I think it does not allow me to push to your branch directly, but here is the fix: [39f3873](https://github.com/macports/macports-ports/commit/39f38737566e0fa87f0ed4897e467ab96d7ab9ab)

Got it, added

**Comment by kencu** [comment]
@barracuda156 
> UPD2. It has been just broken, and fallback is there, just misses a needed macro. I will fix it now.

They just turned down adjusting the macro for older macos 2 weeks ago when the last guy asked.

**Comment by kencu** [comment]
by the way, use ```MAC_OS_X_VERSION_MAX_ALLOWED``` when you are testing SDK features.

use ```MAC_OS_X_VERSION_MIN_REQUIRED``` when you are testing system capabilities.

In this case, you want to know if the needed symbols are available in the system files. So you need to use ```MAC_OS_X_VERSION_MIN_REQUIRED```.

**Comment by barracuda156** [comment]
@kencu Thank you, will change this now.

**Comment by kencu** [comment]
I think Ben kinda doesn't really mind to keep supporting older systems, by the sound of things, but the greater group is dead against it and he's overruled.

**Comment by barracuda156** [comment]
@herbygillot This includes a change advised by Ken above and a non-functional change to style to keep this consistent with PR to `libuv` upstream: https://github.com/macports/macports-ports/commit/2249ce744e3ff065e362762c1384d6fd618015d8 (you can squash on your branch, right?)

**Comment by ryandesign** [comment]
> generate a libuv-legacy port

Please do not create ports with "legacy" in the name unless that is the name of a version or branch used by the developers. Instead, use a version number or branch (e.g. "clang-10" installs version 10.x of clang).

**Comment by ryandesign** [comment]
> To answer your question on adding commits to a PR

It's documented here: https://trac.macports.org/wiki/WorkingWithGit#WorkingwithsomeoneelsespullrequestthroughitsID

File: devel/libuv/files/patch-libuv-legacy-devel.diff [hunk 1]
```diff
@@ -31,7 +31,17 @@ diff --git src/unix/fs.c src/unix/fs.c
...
 --- src/unix/fs.c
 +++ src/unix/fs.c
-@@ -1410,7 +1410,7 @@ static void uv__to_stat(struct stat* src, uv_stat_t* dst) {
+@@ -84,7 +84,8 @@
+ 
+ #if defined(__CYGWIN__) ||                                                    \
+     (defined(__HAIKU__) && B_HAIKU_VERSION < B_HAIKU_VERSION_1_PRE_BETA_5) || \
+-    (defined(__sun) && !defined(__illumos__))
++    (defined(__sun) && !defined(__illumos__)) ||                              \
++    (defined(__APPLE__) && MAC_OS_X_VERSION_MIN_REQUIRED <= 101300)
```
Comment (ryandesign) [inline devel/libuv/files/patch-libuv-legacy-devel.diff:40]: Why "less than or equal to 10.13.0"? Surely you mean either "less than 10.13.0" or "less than 10.14.0" unless you really do want to handle 10.13.0 and earlier differently from 10.13.1 and later.

Comment (barracuda156) [reply devel/libuv/files/patch-libuv-legacy-devel.diff:40]: @ryandesign Are these macros defined also for microversions?

Comment (kencu) [reply devel/libuv/files/patch-libuv-legacy-devel.diff:40]: > @ryandesign Are these macros defined also for microversions?

https://github.com/kencu/macos-sdk/blob/6c1513f5b0667b76e24aaadcad130e90c545f046/MacOSX14.0.sdk/usr/include/AvailabilityVersions.h#L27

Comment (barracuda156) [reply devel/libuv/files/patch-libuv-legacy-devel.diff:40]: Yeah, then this should be changed so that all 10.13.x are treated equally.

**Comment by herbygillot** [comment]
@barracuda156, are we good to merge this? Since this is a new devel port, this should have no impact.

**Comment by barracuda156** [comment]
@herbygillot Let’s address Ryan’s comment above, and then yes. (If you prefer me handling this, I could do it tomorrow, off the machine now.)

**Comment by barracuda156** [comment]
@herbygillot By the way, are there any reasons not to make it a versioned non-conflicting port like it is done for libfmt*?

Not a big deal, just asking. But one practical case is there: Rakudo and friends apparently need a newer `libuv` and cannot be updated for many months, and we cannot depend on a -devel version which conflicts with the standard one.

**Comment by herbygillot** [comment]
> @herbygillot Could you please pick this change, suggested by Ryan above? [barracuda156@c586aed](https://github.com/barracuda156/macports-ports/commit/c586aedf5481710d161755cf2173d881e004b387)

OK, I've added this (combined with the previous legacy patch commit).

> With that, it can be merged. I would prefer having this non-conflicting though, if it is possible and no one objects.

Side-by-side versions of the `libuv` port are, while on the face of it, helpful in a way, I actually feel that it's something we should avoid.  Doing so makes things complicated for other ports in that we can no longer simply say `port:libuv` as a dependency.  There is also the risk of ports getting stuck on older versions of `libuv` for no good reason.

I feel that `libuv` should be mostly backwards-compatible, so there shouldn't be that much trouble in taking the strategy of always seeking to keep the main `libuv` port as up-to-date as possible.

Maybe a possible compromise might be to make the `libuv-devel` port specifically non-conflicting, but multiple versions of `libuv` feels like it would get messy.

**Comment by barracuda156** [comment]
@herbygillot Then let us merge it as-is, then we can try it anything breaks with -devel and whether it fixes Rakudo after all.

(If really needed for Rakudo, I can just switch those three ports to use their own bundled one.)

**Comment by herbygillot** [comment]
Well the desire is if this update looks good for devel, we will look to update the main libuv port to this latest version as well.

**Comment by herbygillot** [comment]
Thank you for your help here @barracuda156

**Pull request merged**